### PR TITLE
Remove numpy as a hard dependency

### PR DESCRIPTION
**Pull request by Fokko**
With Arrow 18.0.0 numpy is not a dependency anymore: https://github.com/apache/arrow/pull/44148

I think it would be good to remove it from PyIceberg.

Closes https://github.com/apache/iceberg-python/issues/1259
Closes https://github.com/apache/iceberg-python/issues/1198

### CODE REVIEW

**Review by kevinjqliu — COMMENTED** [review]
Thanks for the PR, LGTM! 
I've opened #1271 for tracking the performance regression and backporting the pyarrow range function when it becomes available. Do you mind adding it as a comment to the function?

File: pyiceberg/io/pyarrow.py [hunk 1]
```diff
@@ -812,7 +811,15 @@ def _combine_positional_deletes(positional_deletes: List[pa.ChunkedArray], start
         all_chunks = positional_deletes[0]
     else:
         all_chunks = pa.chunked_array(itertools.chain(*[arr.chunks for arr in positional_deletes]))
-    return np.subtract(np.setdiff1d(np.arange(start_index, end_index), all_chunks, assume_unique=False), start_index)
+
+    # Create the full range array with pyarrow
+    full_range = pa.array(range(start_index, end_index))
```
Comment (Fokko) [inline pyiceberg/io/pyarrow.py:816]:
```suggestion
    full_range = pa.array(range(start_index, end_index))
    # When available, replace with Arrow generator to improve performance
    # See https://github.com/apache/iceberg-python/issues/1271 for details
```

**Comment by Fokko** [comment]
@kevinjqliu That's an excellent suggestion, I only created one on the Arrow side, thanks!

**Review by kevinjqliu — APPROVED** [review]
LGTM! bye `numpy`

**Comment by corleyma** [comment]
@kevinjqliu curious what y'all think of my comment here: https://github.com/apache/iceberg-python/issues/1259#issuecomment-2448393527

**Pull request merged**